src/pymor/algorithms/tr.py

```diff
@@ -0,0 +1,292 @@
+    """
+
+    def __init__(self, reductor, initial_guess):
+        self.__auto_init(locals())
+
+        # generate a first rom if none was given
+        self.fom = reductor.fom
+        self.rom = None
+        if reductor._last_rom is None:
+            initial_basis = reductor.fom.solve(initial_guess)
+            reductor.extend_basis(initial_basis)
+            reductor.reduce()
+        self.rom = reductor._last_rom
+        assert self.rom.output_functional is not None, 'Please provide an output functional with your model!'
+
+        # initialize placeholders for extension
+        self.new_reductor = None
+        self.new_rom = None
+
+    def fom_output(self, mu):
+        return self.fom.output(mu)[0, 0]
+
+    def fom_gradient(self, mu):
+        return self.fom.parameters.parse(self.fom.output_d_mu(mu)).to_numpy()
+
+    def rom_output(self, mu):
+        return self.rom.output(mu)[0, 0]
+
+    def estimate_output_error(self, mu):
+        return self.rom.estimate_output_error(mu)
+
+    def extend(self, mu):
+        """Try and extend the current ROM for a new parameter.
+
+        Parameters
+        ----------
+        mu
+            The `Mu` instance for which an extension is computed.
+        """
+        U_h_mu = self.fom.solve(mu)
+        self.new_reductor = deepcopy(self.reductor)
+        try:
+            self.new_reductor.extend_basis(U_h_mu)
+            self.new_rom = self.new_reductor.reduce()
+        except Exception:
+            self.new_reductor = self.reductor
+            self.new_rom = self.rom
+
+    def new_rom_output(self, mu):
+        assert self.new_rom is not None, 'No new ROM found. Did you forget to call surrogate.extend()?'
+        return self.new_rom.output(mu)[0, 0]
+
+    def accept(self):
+        """Accept the new ROM.
+
+        This function is intended to be called after :func:`extend` was called.
+        """
+        assert self.new_rom is not None, 'No new ROM found. Did you forget to call surrogate.extend()?'
+        self.rom = self.new_rom
+        self.reductor = self.new_reductor
+        self.new_rom = None
+        self.new_reductor = None
+
+    def reject(self):
+        """Reject the new ROM.
+
+        This function is intended to be called after :func:`extend` was called.
+        """
+        self.new_rom = None
+        self.new_reductor = None
+
+
+@defaults('beta', 'radius', 'shrink_factor', 'miniter', 'maxiter', 'miniter_subproblem', 'maxiter_subproblem',
+          'tol', 'radius_tol', 'atol', 'tol_sub', 'stagnation_window', 'stagnation_threshold')
+def trust_region(parameter_space, reductor, initial_guess=None, beta=.95, radius=1.,
+                 shrink_factor=.5, miniter=0, maxiter=100, miniter_subproblem=0, maxiter_subproblem=100, tol=1e-6,
+                 radius_tol=.75, atol=1e-6, tol_sub=1e-6, line_search_params=None, stagnation_window=3,
+                 stagnation_threshold=np.inf, return_stages=False, return_subproblem_stages=False):
```

Comment:
I think the default choices of e.g. the tolerances in general should be discussed and evaluated further. Moreover, there should be a different `tol_sub` for the bfgs and for the TR. See the comment in the test.

Reply:
I'm a bit confused by this comment: `tol_sub` is only used in the BFGS part of the algorithm. Do you mean that `atol` and `tol_sub` should differ from one another?

Reply:
I am sorry, my bad. I was referring to `tol` and `tol_sub`, which already differ. So forget about my comment.